### DISCUSSION TITLE
D-latch component that can have default value

### DISCUSSION THREAD

**Discussion by jnoche** [hneemann/Digital#1028] [Ideas]
In an earlier issue (https://github.com/hneemann/Digital/issues/334), you mentioned that at the simulation start phase, gates have different, random propagation times.  That is why, say, the D latch (level-triggered) in the examples\sequential folder (D.dig) would start at a random state.  The solution you proposed was to use memory elements with reset inputs.

Right now, there is a ready-made component called D-Flip-flop (edge-triggered) and its default value can be set (by changing its component properties) without having a reset input.

Is it possible to create a new component called, say, D-Latch (level-triggered) that acts like the D.dig circuit but whose default value can be set (by changing its component properties) without having a reset input?

**Comment by hneemann** [comment]
Yes.
![D](https://user-images.githubusercontent.com/7190563/190847433-86a9f22b-0be7-451c-937a-7286a1b93a6a.png)
But to set the default value to 1 you have to add the reset component to the nor gate at the bottom.
Used as a embedded circuit it has only the inputs D and C.

If you want to adjust the default value via the embedding circuit, a generic circuit must be created, which is possible but not easy to understand and therefore may not be suitable for teaching:

[D.zip](https://github.com/hneemann/Digital/files/9591528/D.zip)

**Reply by jnoche** [reply]
When I try to open the file D.dig in your D.zip file, I get the following error message:

> Error reading a file
Invalid file format
caused by:
---- Debugging information ----
cause-exception : com.thoughtworks.xstream.mapper.CannotResolveClassException
cause-message : romList
class : de.neemann.digital.core.element.ElementAttributes
required-type : de.neemann.digital.core.element.ElementAttributes
converter-type : com.thoughtworks.xstream.converters.reflection.ReflectionConverter
path : /circuit/attributes/entry[2]/romList
line number : 11
class[1] : de.neemann.digital.draw.elements.Circuit
required-type[1] : de.neemann.digital.draw.elements.Circuit
version : not available
>-------------------------------
>caused by: romList

**Reply by jnoche** [reply]
Thank you very much for this. I misunderstood your answer in issue https://github.com/hneemann/Digital/issues/334, I thought the Reset component was an input. Thank you again!

**Reply by hneemann** [reply]
Which Digital version are you using?

**Reply by jnoche** [reply]
OS: Windows 10 OS version: 10.0 JRE: 15.0.1 VM: 15.0.1+9-18 
Digital.jar Build git-Revision: v0.29 Build Time: 2022-02-11 17:13

**Reply by jnoche** [reply]
I put the two files D.dig and Dtest.dig in the examples folder of Digital.

**Reply by jnoche** [reply]
(The info I gave above is what Digital shows me when I select Help>About.  My OS is actually Windows 11.)

**Reply by hneemann** [reply]
Please download the update D.zip and try again.

**Reply by jnoche** [reply]
The new D.dig opens, but the lower NOR gate has only two inputs.  I'm changing it to 3 inputs.

**Reply by jnoche** [reply]
Thank you.  I haven't tested it fully, but it seems to be working now.

**Reply by hneemann** [reply]
> The new D.dig opens, but the lower NOR gate has only two inputs. I'm changing it to 3 inputs.

Damn! Today is really not my best day! :smile:
